velox/functions/sparksql/window/tests/SparkWindowTest.cpp

```diff
@@ -0,0 +1,134 @@
+ */
+#include "velox/common/base/tests/GTestUtils.h"
+#include "velox/functions/lib/window/tests/WindowTestBase.h"
+#include "velox/functions/sparksql/window/WindowFunctionsRegistration.h"
+
+using namespace facebook::velox::exec::test;
+
+namespace facebook::velox::window::test {
+
+namespace {
+
+static const std::vector<std::string> kSparkWindowFunctions = {
+    std::string("nth_value(c0, 1)"),
+    std::string("row_number()")};
+
+struct SparkWindowTestParam {
+  const std::string function;
+  const std::string overClause;
+};
+
+class SparkWindowTestBase : public WindowTestBase {
+ protected:
+  explicit SparkWindowTestBase(const SparkWindowTestParam& testParam)
+      : function_(testParam.function), overClause_(testParam.overClause) {}
+
+  void testWindowFunction(const std::vector<RowVectorPtr>& vectors) {
+    WindowTestBase::testWindowFunction(vectors, function_, {overClause_});
+  }
+
+  void SetUp() override {
+    WindowTestBase::SetUp();
+    WindowTestBase::options_.parseIntegerAsBigint = false;
+    velox::functions::window::sparksql::registerWindowFunctions("");
+  }
+
+  const std::string function_;
+  const std::string overClause_;
+};
+
+std::vector<SparkWindowTestParam> getSparkWindowTestParams() {
+  std::vector<SparkWindowTestParam> params;
+
+  for (auto function : kSparkWindowFunctions) {
+    for (auto overClause : kOverClauses) {
+      params.push_back({function, overClause});
+    }
+  }
+
+  return params;
+}
+
+class SparkWindowTest
+    : public SparkWindowTestBase,
+      public testing::WithParamInterface<SparkWindowTestParam> {
+ public:
+  SparkWindowTest() : SparkWindowTestBase(GetParam()) {}
+};
+
+// Tests all functions with a dataset with uniform distribution of partitions.
+TEST_P(SparkWindowTest, basic) {
+  testWindowFunction({makeSimpleVector(40)});
+}
+
+// Tests all functions with a dataset with all rows in a single partition,
+// but in 2 input vectors.
+TEST_P(SparkWindowTest, singlePartition) {
+  testWindowFunction(
+      {makeSinglePartitionVector(40), makeSinglePartitionVector(50)});
+}
+
+// Tests all functions with a dataset in which all partitions have a single row.
+TEST_P(SparkWindowTest, singleRowPartitions) {
+  testWindowFunction({makeSingleRowPartitionsVector(40)});
+}
+
+// Tests all functions with a dataset with randomly generated data.
+TEST_P(SparkWindowTest, randomInput) {
+  testWindowFunction({makeRandomInputVector(30)});
+}
+
+// Tests nth_value without literal offsets.
+TEST_P(SparkWindowTest, nonLiteralOffset) {
+  vector_size_t size = 20;
+
+  auto vectors = makeRowVector({
+      makeFlatVector<int32_t>(size, [](auto /* row */) { return 1; }),
+      makeFlatVector<int32_t>(size, [](auto row) { return row % 30; }),
+      makeFlatVector<int32_t>(size, [](auto row) { return row % 5; }),
+  });
+
+  std::string overClause = "partition by c0 order by c1";
+  std::string offsetError = "Offset must be literal for spark";
+  assertWindowFunctionError(
```

Comment:
I've been thinking to enhance nth_value() to allow for a column for the second parameter also. From a Velox perspective it would make the implementation more uniform. 

My thinking is that if Spark doesn't allow a column for the second parameter the error should in Spark compiler itself rather than from Window operator runtime.

wdyt ?

Reply:
Yes. I have tried in Gluten with nth_value(c1, c2). The spark compiler will throw exception and don't offload to velox. So this test maybe removed.

Reply:
@JkSelf : Thanks for confirming.